[streaming] [wip] Fault tolerance prototype (for limited case)

mbalassi opened the pull request:
This is the first limited case prototype for streaming fault tolerance. The intention is to provide at least once processing for both states and messages from Kafka. On failure a whole new ExecutionGraph is submitted to the cluster continuing from the last known Kafka offset.

The PR contains three separable tasks:

1. Logic in the JobManager and TaskManager to emit barrier triggering state checkpointing and a new actor to monitor the state (mostly @senorcarbone)
1. New functionality for the record readers handling barriers and state handling on the stream vertex side (mostly @gyfora)
1. Low level Kafka API connector for resetting the offset on failure and storing it as a state (mostly @mbalassi)

Heads-up:

1. @senorcarbone has a bug on the JM side
1. Kafka state is hard-coded in the stream vertex
1. The last Kafka commit is unfinished
1. All three of us are still working on the issue

senorcarbone commented:
hold on, will commit an update very soon. :P

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -471,6 +508,7 @@ void jobVertexInFinalState(ExecutionJobVertex ev) {
 						if (current == JobStatus.FAILING) {
 							if (numberOfRetriesLeft > 0 && transitionState(current, JobStatus.RESTARTING)) {
 								numberOfRetriesLeft--;
+								restart();
```
This seems wrong here. The restart() call must be asynchronous and should respect the delay.

StephanEwen commented on flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/KafkaConsumerIterator.java:
```diff
@@ -0,0 +1,304 @@
+		do {
+			metadata = findLeader(hosts, port, topic, partition);
+			try {
+				Thread.sleep(waitOnEmptyFetch);
```
Why sleep even if the `findLeader(...)` call was successful?

mbalassi replied:
Good catch, no need to sleep there.

StephanEwen commented on flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/KafkaConsumerIterator.java:
```diff
@@ -0,0 +1,304 @@
+			}
+
+			readOffset = messageAndOffset.nextOffset();
+			ByteBuffer payload = messageAndOffset.message().payload();
```
The assumption here is that the Kafka connector reuses its byte buffer, so all returned data has to be copied into a new array?

mbalassi replied:
Indeed it is.

StephanEwen commented on flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/KafkaConsumerIterator.java:
```diff
@@ -0,0 +1,304 @@
+	private void resetFetchResponse(long offset) {
+		FetchRequest req = new FetchRequestBuilder().clientId(clientName)
+				.addFetch(topic, partition, offset, 100000).build();
+		System.out.println(clientName + " " + topic + " " + partition);
```
This sysout call was probably overlooked from the latest debugging session ;-)

mbalassi replied:
Yes, I forgot it.

StephanEwen commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/StreamStateMonitor.scala:
```diff
@@ -0,0 +1,129 @@
+
+object StreamStateMonitor {
+
+  def props(context: ActorContext,executionGraph: ExecutionGraph,
```
This breaks with the Akka semantics: Props creates a properties object that can be used to create an Actor. It does not create the Actor itself.

StephanEwen commented on flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/invokable/operator/co/CoInvokable.java:
```diff
@@ -84,7 +84,16 @@ public void invoke() throws Exception {
 				if (isRunning) {
-					throw e;
+					throw new RuntimeException("Could not read next record due to: "
+							+ StringUtils.stringifyException(e));
```
I think the `stringifyException` is an antipattern. Exceptions should be chained, the stack trace should not go into the message. I would actually go as far as removing `StringUtils.stringifyException`, because it encourages bad use of exceptions.

StephanEwen commented:
Nice first prototype. From my first tries, it seems to work with respect to checkpointing Kafka state in a pipelined (non-blocking) fashion and to reuse it.

Here are a few things I think should be addressed, to make sure that we get high code quality for this sensible piece of the system.

  - This seems to have reproducibly broken the `TaskManagerFailureRecoveryITCase.testRestartWithFailingTaskManager`. Probably through the additional "restart()" call in the execution graph. I am confused why it does not break more.
  
  - This Pull Request "sneaks" in a new job type flag, which marks jobs as "batch" or "streaming". I think this is definitely much needed, and we should make this more prominent.
    
  - Some commits introduce a completely new code style. I have recently written a mail to the mailing list with a discussion about more homogeneity in the project. This is a good example, where there is no good reason not to stick with the common style from the rest of the code. Makes the code much more readable if there is a common style behind it.

  - Related to the previous comment, there are quire a few unnecessary reformattings, which actually take code that was in the original common style, and change it to a different uncommon style.

  - There are many "printStackTrace" command and logging in general is neglected a bit. For a complex mechanism like this, it is super crucial as soon as you need to debug a deployed case.
  
  - Comments are missing big time. This code 
  
  - Messages received by the JobManager/TaskManager which are dedicated for another actor can be "forwarded". That keeps the original sender, which helps in routing answers.
  
  - I would suggest to rename the "StreamStateMonitor" to something along the lines of "StreamCheckpointCoordinator".
  
  - The TaskManager and JobManager have quite a few different messages to handle, would be nice to do more communication endpoint to endpoint (state tracker to source). Not urgent, may be a future refactoring.

  - Along the lines of getting/maintaining a more coherent code base, I would vote to make new classes in the flink-runtime project in Java. Actors via "UntypedActor" are not too bad in Akka/Java.
  
  - I think the best way of desiging a resilinet system is to not make a distinction between the first execution, a recovery execution, a rebalance execution, ... They all should function the same way. There should be no dedicated code paths and flags "onRecovery". The only difference here is that initially, the state is empty, which can be modelled as state as well.

StephanEwen commented:
For the sake of structure, can you also squash a few cleanup commits and prefix all commits consistently with issue numbers and component (https://cwiki.apache.org/confluence/display/FLINK/Apache+Flink+development+guidelines)

senorcarbone commented:
thanks for the review Stephan! The synchronous call to executionGraph restart was a very sneaky thing, I committed it by mistake while trying some things out and it's most probably why the test case fails, my bad. I addressed some of the quick changes  you mentioned already to my last squashed commit.

mbalassi commented:
Thanks for the review, @StephanEwen. We have addressed the issues you have mentioned and squashed the commits a bit. Hopefully the PR is ready to  merge.

mbalassi commented:
A `setNumberOfExecutionRetries(int)` method is really needed in the `StreamExecutionEnvironment`. I'm adding it to a separate upcoming PR working on [1].

[1] https://issues.apache.org/jira/browse/FLINK-1669

StephanEwen commented:
Looks good. I will make a final pass and merge it...

StephanEwen commented:
Testing so far indicates all issues are fixed. Rebasing worked without problems...

StephanEwen commented:
Last check on Travis before merging.

There are a couple of followup changes I would propose, let's start a separate thread about this on the mailing list.
